### DISCUSSION TITLE
Debug Symbols Ignored (linux-x64)

### DISCUSSION THREAD

**Discussion by joshm-cognitiv** [microsoft/vcpkg#33560] [Q&A] [answered]
Hello,
I'm trying to run a debug version of my project linked against vcpkg generated deps and cmake seems to be ignoring the debug libs even though they exist. I'm using the default x64-linux triplet for my build with CMAKE_BUILD_TYPE=Debug.

The vcpkg_installed directory has the following layout
```
x64-linux/debug/lib (debug)
x64-linux/lib (release bins)
```
cmake doesn't even seem to know about debug/lib and only links with x64-linux. Is there some special flag I need for getting cmake/vcpkg to use debug/lib? fwiw I tried removing x64-linux/lib and things blowup... which is to be expected.

**Comment by Osyotr** [comment]
Any example? The issue you've described sounds like an incorrect usage of cmake (i.e. setting `CMAKE_TOOLCHAIN_FILE` after `project()` call, hardcoding paths to cmake configs, not using imported targets etc.)

**Comment by joshm-cognitiv** [comment]
So there is a root cmake file that has the following
```
cmake_minimum_required(VERSION 3.20)

# custom triplets (not used for question)
set(VCPKG_OVERLAY_TRIPLETS "${CMAKE_SOURCE_DIR}/triplets")

if(DEFINED ENV{VCPKG_ROOT})
    set(CMAKE_TOOLCHAIN_FILE "$ENV{VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake" CACHE STRING "Vcpkg toolchain file")
else()
    set(CMAKE_TOOLCHAIN_FILE "./vcpkg/scripts/buildsystems/vcpkg.cmake" CACHE STRING "Vcpkg toolchain file")
endif()

set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -fPIC -fno-omit-frame-pointer")
set(CMAKE_EXPORT_COMPILE_COMMANDS 1)
set(CMAKE_CXX_STANDARD "17")

project(proj_name CXX)

set(CMAKE_MODULE_PATH "${PROJECT_SOURCE_DIR}/cmake;${CMAKE_MODULE_PATH}")

# Use for colored compiler output with Ninja build tool
option (ENABLE_COLORED_OUTPUT "Always produce ANSI-colored output (GNU/Clang only)." OFF)
if (${ENABLE_COLORED_OUTPUT})
    if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
       message("Enabling colored compiler output for gcc")
       add_compile_options (-fdiagnostics-color=always)
    elseif ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
       message("Enabling colored compiler output for color")
       add_compile_options (-fcolor-diagnostics)
    else()
      message("Ignoring colored compiler output, unknown compiler id ${CMAKE_CXX_COMPILER_ID}")
    endif ()
endif ()

# some example finds.
find_package(spdlog REQUIRED)
find_package(fmt REQUIRED)
find_package(Catch2 3 REQUIRED)

# rest of project building...
```

We use a git submodule containing vcpkg with vcpkg-configuration.json pointing to it with 'built-in'. Projects are generally initialized with
` cmake -B . -S ..  -DCMAKE_BUILD_TYPE=Debug`

If there is something else I can provide please let me know. We also currently build against an August release of vcpkg repo.

**Reply by Osyotr** [reply]
Looks OK. What makes you think that CMake links wrong libs?

**Reply by joshm-cognitiv** [reply]
gdb does not load the symbols for all the dependencies built by vcpkg. I can confirm this by remapping x64-linux/debug/lib to x64-linux/lib. It does actually seem to load some symbols for a set of the libraries, but others it does not. Does this mean its possibly some cmake issue in a dependency?

**Reply by Osyotr** [reply]
x64-linux is a static triplet, all symbols are embedded into your executable/library.
If there are libraries that can't be built statically, then you might want to create custom triplet with  `VCPKG_FIXUP_ELF_RPATH` in it.

**Reply by joshm-cognitiv** [reply]
Well they are all statically linked. I'll dig some more and get back to you tomorrow. Thanks for the help!

**Comment by dg0yt** [comment]
Check that your own find modules properly handle vcpkg's debug libs, and that no system lib is selected when you expect a vcpkg lib. It can help to inspect `CMakeCache.txt`.

**Reply by joshm-cognitiv** [reply, marked as answer]
Ok, I tracked down the issue. Turns out the libraries I didn't have symbols for were not configuring their ports correctly to copy over cmake's debug configuration of the targets (specifically ` vcpkg_cmake_config_fixup`). Once this was fixed, everything worked as expected. Thanks for the help, sorry for any confusion this may have caused.
